### ISSUE TITLE
If the IMAP4 server receives a SEARCH command with an illegal term, it will raise an unhandled error and never respond to the client

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4080]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4080|
|Type|defect|
|Created|2009-11-11 20:52:14Z|
|Branch|https://github.com/twisted/twisted/tree/illegal-search-4080|

This is because search is implemented with dynamic dispatch but without considering the case where there is no `search_` method corresponding to the query term.

This is related to [#1977](https://github.com/twisted/twisted/issues/1977) which encountered the problem when it misinterpreted a message sequence set as an illegal term.

Attachments:

* [twisted-4080.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f5c/f5ca8c8552c95526a48c9a4b7db15fa4e86c11f8/4dab50a67b6049306045083514e3bd2c62b503ee.patch) (1653 bytes) - added by wulczer on 2010-03-21 17:12:28Z - 
* [imap_illegal_search_4080.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f5c/f5ca8c8552c95526a48c9a4b7db15fa4e86c11f8/fe6965ba4a49f697106d29ad131a1cfeb08cdc72.patch) (6634 bytes) - added by danmil on 2012-07-06 15:55:10Z - 
* [imap_illegal_search_4080_2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f5c/f5ca8c8552c95526a48c9a4b7db15fa4e86c11f8/c2c4643a83b313ee7ff1d8690126c983166d9c60.patch) (7308 bytes) - added by danmil on 2012-07-30 18:27:36Z - Second attempt at patch, incorporating exarkun's comments.

<details><summary>Searchable metadata</summary>

```
trac-id__4080 4080
type__defect defect
reporter__exarkun exarkun
priority__low low
milestone__ 
branch__branches_illegal_search_4080 branches/illegal-search-4080
branch_author__exarkun__jesstess exarkun, jesstess
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1257972734000000 1257972734000000
changetime__1344282903000000 1344282903000000
version__None None
owner__exarkun exarkun
cc__jesstess cc__danmil
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="wulczer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|wulczer commented|
|-|-|

Proposed patch.

It makes a SEARCH FOO command not to fail in the server, but to not return any results.

This seems to differ from the original intention of the code, which was missing an argument to getattr, but just adding that argument would make SEARCH FOO commands to return all messages in the mailbox.

IMHO not returning anything makes more sense than returning everything (even more sensible would be returning an error to the client, but that'd require some more work.

**Comment by twisted-trac** [comment]
|<img alt="wulczer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|wulczer commented|
|-|-|

Oh, that patch applies over the one from [#2278](https://github.com/twisted/twisted/issues/2278).

**Comment by twisted-trac** [comment]
|<img alt="wulczer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|wulczer removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_5"></a>|@jesstess commented|
|-|-|

(In [28796]) Branching to 'imap-illegal-search-4080'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_6"></a>|@jesstess set owner to wulczer|
|-|-|

wulczer: before I apply the patch on this branch - in my brief perusal of RFC 3501, "6.4.4.SEARCH Command" it says

   Result:     OK - search completed
               NO - search error: can't search that [CHARSET] or
                    criteria
               BAD - command unknown or arguments invalid

which suggests that one of NO or BAD should be the result. Can you comment on why not returning anything as in your patch is more correct or update the patch?

Thanks for working on these IMAP tickets!

**Comment by twisted-trac** [comment]
|<img alt="wulczer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|wulczer commented|
|-|-|

Replying to [jesstess](#note_6):

> which suggests that one of NO or BAD should be the result. Can you comment on why not returning anything as in your patch is more correct or update the patch?

Because it meant less work :-)

I'll try to come up with a more complete fix.

**Comment by twisted-trac** [comment]
|<img alt="danmil's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|danmil removed owner|
|-|-|

This properly fixes the IMAP server so that, when it receives a SEARCH command with an illegal term, it returns an IMAP error code to the client, and logs an error server side.

I added unit tests and fixes for both 'manual' search (where the server loads all messages from a mailbox and runs the search itself), and also search of a mailbox which implements ISearchableMailbox.

A key problem turned out to be that the do_SEARCH method was stacking together the callback and errback via addCallback**s**, which meant that error backs firing out of the first **callback** were not getting caught.  I switched that so that the errBack is attached to the first callback, then had the actual search code raise an exception.

Implementers of ISearchableMailbox are now responsible for raising IllegalQueryError if they find an error in the search criteria (before, there was nothing specified about what they could or should do in that case -- in fact, there was no way for an error to be signalled back to the client, I believe).  I added a 'raises' description to the interface to make that clear, and mocked it up in the test.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to danmil<br>@exarkun set status to `new`|
|-|-|

Thanks for working on Twisted's IMAP4 support.  It's really great to see improvements in this area.

  1. twisted/mail/imap4.py
    1. The old code in `_singleSearchStep` looked like it was _trying_ to handle this case, at least.  It tested `f` (the result of the `getattr`) against `None`.  But it forgot to pass `None` as the third argument to the `getattr` call, so it didn't actually catch the case where no such search method existed.  The `if f is not None:` spelling is preferable to the `if not f:` spelling: the boolean truth of `f` isn't of much interest, it's whether the fallback/default case of `getattr` that was hit that's important.
    1. There's a bunch of new trailing whitespace added in the patch, which we try to keep out of Twisted.
  1. twisted/mail/test/test_imap.py
    1. There's a helper in trial, `TestCase.assertFailure`, that I think will simplify these new tests.  It should be usable to replace the `searchSucceeded` callback.
    1. There are some trivial but critical mistakes in the two new test methods: use of the undefined name `result`, for example.  Since the tests pass anyway, this reveals a big problem with them: even in the presence of random exceptions, they don't fail.  I suggest applying just the tests to a pristine Twisted trunk working copy and then working on them until they fail in the anticipated way.  Then you should be able to make them pass by applying the imap4.py parts of the patch.  This should ensure the tests are providing good coverage of the functionality.
    1. Also, boringly, make sure everything has a docstring, name test methods like `test_foo`, and also clean up the trailing whitespace in the new test methods.
  1. It'd also be great if you could supply a [ReviewProcess#Newsfiles news fragment] for this change.

Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="danmil's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|danmil set owner to @exarkun|
|-|-|


Great comments, thanks for the detailed response.  The newly attached patch applies as much of this as possible, details below:

 1- First off, as to the big, bad gotcha -- the tests have a syntax error but don't fail

Short version: argh.  Longer version: I did actually develop these tests failure first, but that one line got caught by my not fully understanding the combo of error backs + the trial unit test context.  Specifically, I copied in some other code that attached a _cbStopClient errBack, which was silently eating some (but not all, AFAICT) errors from my actual error back code.  I've replaced that with an errBack which reraises errors, and things are much better.

I belive I've now got that all properly sorted out.  I did run the tests against a clean tree -- as expected, one of them (test_searchInvalidCriteria) fails.  The other does not -- now that I look carefully, I see that, if an ISearchableMailbox had raised IllegalQueryError, the right thing was happening (the issue with where the errback got attached is not crippling in that case, as it was for the other).  I'd still like to leave the slight adjustment to the first branch in ```do_SEARCH```, because it properly handles errors if there's an exception in ```__cbSearch```, which the old code did not.


 2- assertFailure

I tried to get this to work, but AFAICT, whenever I used it, it was 'handling' the errors before my more-specific error handling code could get at it.  Which wasn't actually okay -- I need my specific error handling code to fire so I can verify both the error which gets logged and the results sent back to the client.  This might be related to the fact that the test case is running both the client and the server, somehow attached to that deferred.  But, in any event, whenever I used assertFailure, I couldn't get my error code to run, and I was seeing the test fail in various ways.


 3- test_foo

I did **not** rename testInvalidTerm to test_invalidTerm because all of the other test methods in FetchSearchStoreTestCase are, e.g. testSearch, testUIDSearch, etc.  If you'd like, I'd be happy to rename all of those.


 4- Trailing whitespace, docstring, topfiles, not f -> f is None

Got 'em all, I think.

Finally, I'm assigning back to you -- I'm not sure if that's the right review protocol, if not, my apologies.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [35071]) Branching to 'illegal-search-4080'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [35072]) Apply imap_illegal_search_4080_2.patch 

refs [#4080](https://github.com/twisted/twisted/issues/4080)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [35073]) Switch to using TestCase.assertFailure instead of a custom implementation of the same

refs [#4080](https://github.com/twisted/twisted/issues/4080)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

> I tried to get [`assertFailure`] to work, but AFAICT, whenever I used it, it was 'handling' the errors before my more-specific error handling code could get at it.

It handles them by verifying they occur, then makes the exception instance the result of the `Deferred`.  I made the change in r35073 so you can see how it's done.

> I did not rename testInvalidTerm to test_invalidTerm because all of the other test methods in FetchSearchStoreTestCase are, e.g. testSearch, testUIDSearch, etc. If you'd like, I'd be happy to rename all of those.

I'm not a fan of mass renamings.  Given that, the only way we'll get the code updated to the new standard is if new code follows the new standard, even if it disagrees with local convention.  Disagreeing with local convention isn't super awesome either, but I think it's preferable in this case.

> I'd still like to leave the slight adjustment to the first branch in do_SEARCH, because it properly handles errors if there's an exception in __cbSearch, which the old code did not.

That's okay, I suppose.  In practice, there will probably never be an exception from `__cbSearch` though.  And if it is possible to have one, there should be a unit test covering it.  Just a couple things to keep in mind in case the situation arises again. :)

The changes look good to me now.  If the [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/illegal-search-4080) look good, I'll merge.  Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `closed`|
|-|-|

(In [35079]) Merge illegal-search-4080

Author: danmil
Reviewer: exarkun
Fixes: [#4080](https://github.com/twisted/twisted/issues/4080)

Change `IMAP4Server` to handle errors from the search implementation and send back a proper
error response to the client when they occur.
